packages/password/index.js

```diff
@@ -0,0 +1,132 @@
+const {generatePasswordFromInput} = require('./lib/apple.password')
+const {ParserError} = require('./lib/rules-parser')
+const {constants} = require('./lib/constants')
+
+/**
+ * @typedef {{
+ *   domain?: string | null | undefined;
+ *   input?: string | null | undefined;
+ *   rules?: RulesFormat | null | undefined
```

Comment:
Yes, this one is about ergonomics - I discovered usability issues when I built out my prototype.

For example, a common use-case will be to use a input field's `passwordrules` OR the current domain, which in practice looks like this with my API.

```javascript
const pw = password.generate({ 
   input: inputElement.getAttribute("passwordrules"),
   domain: window.location.hostname,
})
```

So the idea is that a consumer may provide `inputElement.getAttribute("passwordrules")` -> which may return `null`, *or* you may be providing the input from a stored value that could be `undefined`. Same thing applies for `domain`. 

So the design of my API allows you to pass optional values in a single call and always get a password back.

The alternative would lead to a more complex API where we would need to propagate errors back to the user and force them to try alternative methods etc.

My design also keeps error handling internal and ensures that the order of precedence is preserved (eg: the internals of the library chooses which to try first (input, domain or default)